Fix compat issue with Type.IsAssignableFrom

steveharter opened the pull request:
Adds tests for  https://github.com/dotnet/coreclr/issues/20154 although they will fail until the CLR [change ](https://github.com/dotnet/coreclr/pull/20252)propagates to corefx.

jkotas commented on src/System.Reflection/tests/MockDerivedType.cs:
```diff
@@ -0,0 +1,104 @@
+
+namespace System.Reflection.Tests
+{
+    internal class MockDerivedType : Type
```
There is `MockType` under Common already. Inherit from that so that you do not have to list overrides from all properties from it?

Once you do that, it should be pretty small. It can be in the same file as the test that it is testing.

jkotas commented on src/Common/src/CoreLib/System/Type.Helpers.cs:
```diff
@@ -342,7 +342,7 @@ public virtual bool IsAssignableFrom(Type c)
             // whose UnderlyingSystemType are runtime implemented. 
             Type toType = this.UnderlyingSystemType;
-            if (toType.IsRuntimeImplemented())
+            if (toType != null && toType.IsRuntimeImplemented())
```
Nit: This can be `toType?.IsRuntimeImplemented()`

steveharter replied:
Hmm since the bool is a primitive, using ? means the expression is now bool? requiring this:
`toType?.IsRuntimeImplemented() == true`
which may or may not be more readable than
`toType != null && toType.IsRuntimeImplemented()`

jkotas replied:
Ok. `toType != null && toType.IsRuntimeImplemented()` looks better then.

jkotas commented on src/System.Reflection/tests/TypeDerivedTests.cs:
```diff
@@ -0,0 +1,24 @@
+        {
+            Type compareType = typeof(TypeDerivedTests);
+            var testType = new MockDerivedType();
+            Assert.Null(testType.UnderlyingSystemType);
```
I think this can be added as more `InlineData` to the existing IsAssignableFrom test in TypeInfoTests.cs. I do not think we need a special test for this.

steveharter replied:
The IsAssignableFrom test in TypeInfoTests specifically uses `TypeInfo` (e.g. type.GetTypeInfo()) so not a fit for this.

jkotas replied:
The `TypeInfo` tests are really testing `Type`. They were written in the netstandard1.x days when `Type` had very few methods, and all interesting methods were on `TypeInfo`. It is not case anymore.

It is fine to keep it here if you prefer.

jkotas commented:
> The fix (in Type.Helpers.cs) is in a common location shared with coreclr, so the fix will need to propagate to coreclr repo

In situations like these, we typically check-in the fix to CoreCLR first and then check in the test to CoreFX once the change propagates.

The problem with checking fixes like these to CoreFX first is that you will often find that the fix does not build once propagated to CoreCLR.

steveharter commented:
@jkotas 
> In situations like these, we typically check-in the fix to CoreCLR first and then check in the test to CoreFX once the change propagates.

I'll send out the CoreCLR PR soon, and then a new commit here addressing feedback once we get a CoreCLR build. Thanks

jkotas commented on src/System.Reflection/tests/TypeDerivedTests.cs:
```diff
@@ -0,0 +1,29 @@
+    internal class MockDerivedType : BaseMockType
+    {
+        public override Type UnderlyingSystemType => null;
+        protected override TypeAttributes GetAttributeFlagsImpl() => new TypeAttributes();
```
It may be better to use `TypeAttributes.Class` instead of `new TypeAttributes()`.

jkotas commented on src/Common/tests/System/MockType.cs:
```diff
@@ -74,7 +74,6 @@ internal class MockType : Type
         public override MemberInfo[] GetMember(string name, BindingFlags bindingAttr) => throw Unexpected;
         public override MemberInfo[] GetMember(string name, MemberTypes type, BindingFlags bindingAttr) => throw Unexpected;
         protected override TypeCode GetTypeCodeImpl() => throw Unexpected;
-        public override bool IsAssignableFrom(Type c) => throw Unexpected;
```
Are the other tests using the MockType broken if this method stays as is in the MockType?

steveharter replied:
I just verified an no - I could just leave one MockType if we don't care about slight pollution

jkotas replied:
I think no BaseMockType would look better. Type has number of non-virtual methods, so there is some impurity by design. If there is ever a test that actually wants IsAssignableFrom to throw, it can have a specialized Mock type that is just for the given test.

steveharter replied:
Ok

jkotas commented on src/System.Reflection/tests/TypeDerivedTests.cs:
```diff
@@ -0,0 +1,29 @@
+
+namespace System.Reflection.Tests
+{
+    internal class MockDerivedType : MockType
```
Nit: A better name for this may be `TypeWithNullUnderlyingSystemType`. It is a very specific mock type; not an arbitrary one.

jkotas commented:
@dotnet-bot test this please

jkotas commented:
The new test is failing:

```
 at System.MockType.get_IsGenericParameter() in D:\j\workspace\windows-TGrou---f8ac6754\src\Common\tests\System\MockType.cs:line 82
   at System.Type.IsAssignableFrom(Type c) in E:\A\_work\290\s\src\System.Private.CoreLib\shared\System\Type.Helpers.cs:line 356
   at System.Reflection.Tests.TypeDerivedTests.IsAssignableFrom_NullUnderlyingSystemType() in D:\j\workspace\windows-TGrou---f8ac6754\src\System.Reflection\tests\TypeDerivedTests.cs:line 25
```

steveharter commented:
>> The new test is failing:
Should be fixed. The MockType base class overrode an additional member than the previous test class
